File: TypeScript/src/frogger/frogger.ts
import { Road } from "./road";

/**
 * Refactor Task 1, 2: Frogger
 *
 * @author Zishen Wen (F22), Deyuan Chen (S22), Duan Liang (F23)
 */
class Frogger {

  // Field for Task 1.
	private road: Road;
  private position: number;

  // Field for Task 2. Anything to add?
  private records: Records;
  private firstName: string; 
  private lastName: string; 
  private phoneNumber: string;
  private zipCode: string;
  private state: string;
  private gender: string;
  
  constructor(road: Road, position: number, records: Records, firstName: string, 
    lastName: string, phoneNumber: string, zipCode: string, state: string, gender: string) {
      this.road = road;
      this.position = position;
      this.records = records;
      this.firstName = firstName;
      this.lastName = lastName;
      this.phoneNumber = phoneNumber;
      this.zipCode = zipCode;
      this.state = state;
      this.gender = gender;
  }

	/**
   * Moves Frogger.
   * @param forward true is move forward, else false.
   * @return true if move successful, else false.
   */
  public move(forward: boolean): boolean {
      let nextPosition: number = this.position + (forward ? 1 : -1);
      if (!this.isValid(nextPosition) || this.isOccupied(nextPosition)) {
          return false;
      }
      this.position = nextPosition;
      return true;
  }

	// TODO: Do you notice any issues here?
  public isOccupied(position: number): boolean {
      let occupied: boolean[] = this.road.getOccupied();
      return occupied[position];
  }

  public isValid(position: number): boolean {
      if (position < 0) return false;
      let occupied: boolean[] = this.road.getOccupied();
      return position < occupied.length;
  }

  /**
   * Records Frogger to the list of records.
   * 
   * @returns true if record successful, else false.
   */
  public recordMyself(): boolean {
    // This is an annoying call...
    let success: boolean = this.records.addRecord(this.firstName, this.lastName, this.phoneNumber, this.zipCode, this.state, this.gender)
    return success
  }
}

File: Java/src/main/java/frogger/Frogger.java
package frogger;

/**
 * Refactor Task 1 & 2: Frogger
 *
 * @author Zishen Wen (F22), Deyuan Chen (S22), Duan Liang (F23)
 */
public class Frogger {

    // Field for task 1.
    private final Road road;
    private int position;
    
    // Field for task 2. Anything to add?
    private final Records records;
    private String firstName, lastName, phoneNumber, zipCode, state, gender;

    // We won't pass in all the identity strings above to avoid messy code. Can we pass in something else..?
    public Frogger(Road road, int position, Records records) {
        this.road = road;
        this.position = position;
        this.records = records;
    }

    /**
     * Moves Frogger.
     *
     * @param forward true is move forward, else false.
     * @return true if move successful, else false.
     */
    public boolean move(boolean forward) {
        int nextPosition = this.position + (forward ? 1 : -1);
        if (!isValid(nextPosition) || isOccupied(nextPosition)) {
            return false;
        }
        this.position = nextPosition;
        return true;
    }

    // TODO: Do you notice any issues here?
    public boolean isOccupied(int position) {
        boolean[] occupied = this.road.getOccupied();
        return occupied[position];
    }
    
    public boolean isValid(int position) {
        if (position < 0) return false;
        boolean[] occupied = this.road.getOccupied();
        return position < occupied.length;
    }

    /**
     * Records Frogger to the list of records.
     * 
     * @return true if record successful, else false.
     */
    public boolean recordMyself() {
      boolean success = records.addRecord(firstName, lastName, phoneNumber, zipCode, state, gender);
      return success;
    }

}
